test/e2e/reporters/progress.go

```diff
@@ -0,0 +1,148 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reporters
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/config"
+	"github.com/onsi/ginkgo/types"
+	"k8s.io/klog"
+)
+
+// ProgressReporter is a ginkgo reporter which tracks the total number of tests to be run/passed/failed/skipped.
+// As new tests are completed it updates the values and prints them to stdout and optionally, sends the updates
+// to the configured URL.
+type ProgressReporter struct {
+	LastMsg string `json:"msg"`
+
+	NumToBeRun   int `json:"total"`
+	NumCompleted int `json:"completed"`
+	NumSkipped   int `json:"skipped"`
+	NumFailed    int `json:"failed"`
+
+	Failures []string `json:"failures,omitempty"`
+
+	progressURL string
+	client      *http.Client
+}
+
+// NewProgressReporter returns a progress reporter which posts updates to the given URL.
+func NewProgressReporter(progressUpdateURL string) *ProgressReporter {
+	c := &http.Client{
+		Timeout: time.Second * 10,
+	}
+	return &ProgressReporter{
+		Failures:    []string{},
+		progressURL: progressUpdateURL,
+		client:      c,
+	}
+}
+
+// SpecSuiteWillBegin is invoked by ginkgo when the suite is about to start and is the first point in which we can
+// antipate the number of tests which will be run.
+func (reporter *ProgressReporter) SpecSuiteWillBegin(cfg config.GinkgoConfigType, summary *types.SuiteSummary) {
+	reporter.NumToBeRun = summary.NumberOfSpecsThatWillBeRun
+	reporter.LastMsg = "Test Suite starting"
+	reporter.sendUpdates()
+}
+
+// SpecSuiteDidEnd is the last method invoked by Ginkgo after all the specs are run.
+func (reporter *ProgressReporter) SpecSuiteDidEnd(summary *types.SuiteSummary) {
+	reporter.LastMsg = "Test Suite completed"
+	reporter.sendUpdates()
+}
+
+// SpecDidComplete is invoked by Ginkgo each time a spec is completed (including skipped specs).
+func (reporter *ProgressReporter) SpecDidComplete(specSummary *types.SpecSummary) {
+	testname := strings.Join(specSummary.ComponentTexts[1:], " ")
+	switch specSummary.State {
+	case types.SpecStateFailed:
+		if len(specSummary.ComponentTexts) > 0 {
+			reporter.Failures = append(reporter.Failures, testname)
+		} else {
+			reporter.Failures = append(reporter.Failures, "Unknown test name")
+		}
+		reporter.NumFailed++
+		reporter.LastMsg = fmt.Sprintf("FAILED %v", testname)
+	case types.SpecStatePassed:
+		reporter.NumCompleted++
+		reporter.LastMsg = fmt.Sprintf("PASSED %v", testname)
+	case types.SpecStateSkipped:
+		reporter.NumSkipped++
+		return
+	default:
+		return
```

Comment:
just a question: This `default` case is unexpected, right?

Reply:
Correct; we are just parsing for success/failure/skip. There are other types but they aren't really applicable at this time for how we're using this (https://github.com/onsi/ginkgo/blob/000d317071340972710dc1da9ae3c2c016a7a560/types/types_test.go#L10-L16)